### PR TITLE
multiquic support in solana_streamer -- rebase of #634 to latest master

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem
We have found in production that the quic streamer endpoint is overwhelmed with connections and txn packets that cause large number of connection timeout or send transaction time out. With multiple endpoints, multiple threads can be utilized to ingest the input packets.

#### Summary of Changes

* tpu: use multiple quic endpoints

* cluster-info: manage port range by hand...

* local-cluster: keep udp tpu socket around for tests

* First set the number of endpoint to 1 -- pending more test results in the latest v1.18 on skip rage problem.

Fixes #

### CODE REVIEW

File: local-cluster/src/local_cluster.rs [hunk 1]
```diff
@@ -337,7 +337,7 @@ impl LocalCluster {
             socket_addr_space,
             DEFAULT_TPU_USE_QUIC,
             DEFAULT_TPU_CONNECTION_POOL_SIZE,
-            DEFAULT_TPU_ENABLE_UDP,
+            true,
```
Comment (alessandrod) [inline local-cluster/src/local_cluster.rs:342]: I remember I did this change, but I don't remember why 😅 We should add a
comment here

**Comment by alessandrod** [comment]
I've re-re-re-reviewed this. It looks great (😏), except for the manual range thing, which I would try to undo.

See this commit: https://github.com/solana-labs/solana/pull/1149/commits/fb0eea2df082bf9a47bd1edfefefaf626439012e. I think if we revert this - so we don't set reuseport=true in multi_bind_in_range - it should work (tm).

I don't think that reverting that commit actually breaks anything. In most other places we don't set reuseport=true, so WSL will break there anyway. I'd just revert it.

File: net-utils/src/lib.rs [hunk 2]
```diff
@@ -384,29 +384,63 @@ pub fn is_host_port(string: String) -> Result<(), String> {
...
     let sock = Socket::new(Domain::IPV4, Type::DGRAM, None)?;
-    let sock_fd = sock.as_fd();
 
+    // best effort, i.e. ignore setsockopt() errors, we'll get the failure in caller
     if reuseaddr {
-        // best effort, i.e. ignore errors here, we'll get the failure in caller
-        setsockopt(&sock_fd, ReusePort, &true).ok();
-        setsockopt(&sock_fd, ReuseAddr, &true).ok();
+        setsockopt(&sock, ReuseAddr, &true).ok();
+        reuseport = true;
```
Comment (alessandrod) [inline net-utils/src/lib.rs]: I think we should also remove this

File: net-utils/src/lib.rs [hunk 3]
```diff
@@ -484,8 +527,12 @@ pub fn multi_bind_in_range(
             port
         }; // drop the probe, port should be available... briefly.
 
+        let config = SocketConfig {
+            reuseaddr: true,
+            reuseport: true,
```
Comment (alessandrod) [inline net-utils/src/lib.rs]: This should be reuseport: false

File: local-cluster/src/local_cluster.rs (hunk 1 above)
Comment (lijunwangs) [reply local-cluster/src/local_cluster.rs:342]: Added comments

File: net-utils/src/lib.rs (hunk 2 above)
Comment (lijunwangs) [reply net-utils/src/lib.rs]: Done

File: net-utils/src/lib.rs (hunk 3 above)
Comment (lijunwangs) [reply net-utils/src/lib.rs]: I think this comment is no longer relevant, can you check again?

**Review by alessandrod — APPROVED** [review]

File: net-utils/src/lib.rs (hunk 3 above)
Comment (alessandrod) [reply net-utils/src/lib.rs]: I mean, I think this should still be false, but there are other places too in net-utils that set reuseport=true for no reason. We don't have to fix it all in this PR, it's pre existing bugs, they can be fixed in followup PRs.

**Pull request merged**